File: Source/Fantasy.Jobs/Scheduling/ScheduleItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Runtime.Serialization;
using Fantasy.XSerialization;
using System.Xml.Linq;
using System.Xml.XPath;
using System.Xml;

namespace Fantasy.Jobs.Scheduling
{
    [DataContract,
     KnownType(typeof(TimeTrigger)),
     KnownType(typeof(DailyTrigger)),
     KnownType(typeof(WeeklyTrigger)),
     KnownType(typeof(MonthlyTrigger)),
     KnownType(typeof(MonthlyDOWTrigger)),
     KnownType(typeof(TemplateAction)),
     KnownType(typeof(InlineAction)),
     KnownType(typeof(CustomAction))
    ]
    [XSerializable("schedule", NamespaceUri = Consts.ScheduleNamespaceURI)]
    public class ScheduleItem : IXSerializable
    {
        public ScheduleItem()
        {
            
        }

        [DataMember]
        [XAttribute("name")]
        public string Name { get; set; }

        [DataMember]
        [XAttribute("author")]
        public string Author { get; set; }

        [DataMember]
        [XAttribute("description")]
        public string Description { get; set; }

        [DataMember]
        [XAttribute("enabled")]
        public bool Enabled { get; set; }

        [DataMember]
        public Trigger Trigger { get; set; }

        [DataMember]
        public Action Action { get; set; }

        [DataMember]
        [XAttribute("priority")]
        public int Priority { get; set; }

        [DataMember]
        [XAttribute("restartOnFailure")]
        public Restart RestartOnFailure { get; set; }

        [DataMember]
        [XArray(Name = "requiredUncPath"),
        XArrayItem(Name = "unc", Type = typeof(UncPath))]
        public UncPath[] RequiredUncPths { get; set; }

        [DataMember]
        [XAttribute("runOnlyIfIdle")]
        public bool RunOnlyIfIdle { get; set; }

        [DataMember]
        [XAttribute("startWhenAvailable")]
        public bool StartWhenAvailable { get; set; }

        [DataMember]
        public string CustomParams { get; set; }


        [DataMember]
        [XAttribute("multipleInstances")]
        public InstancesPolicy MultipleInstance { get; set; }


        [DataMember]
        [XAttribute("expired")]
        public bool Expired { get; set; }

        [DataMember]
        [XAttribute("deleteAfterExpired")]
        public bool DeleteAfterExpired { get; set; }

        #region IXSerializable Members

        void IXSerializable.Load(IServiceProvider context, XElement element)
        {
            XHelper.Default.LoadByXAttributes(context, element, this);
           
            XmlNamespaceManager nm = new XmlNamespaceManager(new NameTable());
            nm.AddNamespace("s", Consts.ScheduleNamespaceURI);

            XElement customParams = element.XPathSelectElement("s:custom", nm);
            if (customParams != null)
            {
                this.CustomParams = customParams.ToString();
            }


            XElement triggerElement = element.XPathSelectElement("s:trigger", nm);
            if (triggerElement != null)
            {
                Type type;
                TriggerType triggerType = (TriggerType)Enum.Parse(typeof(TriggerType), (string)triggerElement.Attribute("type"));
                switch (triggerType)
                {
                    case TriggerType.Time:
                        type = typeof(TimeTrigger);
                        break;
                    case TriggerType.Daily:
                        type = typeof(DailyTrigger);
                        break;
                    case TriggerType.Weekly:
                        type = typeof(WeeklyTrigger);
                        break;
                    case TriggerType.Monthly:
                        type = typeof(MonthlyTrigger);
                        break;
                    case TriggerType.MonthlyDayOfWeek:
                        type = typeof(MonthlyDOWTrigger);
                        break;
                    default:
                        throw new Exception("Absurd!");
                }
                XSerializer ser = new XSerializer(type);
                ser.Context = context;
                this.Trigger = (Trigger)ser.Deserialize(triggerElement);

            }

            XElement actionElement = element.XPathSelectElement("s:action", nm);
            if (actionElement != null)
            {
                Type type;
                ActionType actionType = (ActionType)Enum.Parse(typeof(ActionType), (string)actionElement.Attribute("type"));
                switch (actionType)
                {
                    case ActionType.Template:
                        type = typeof(TemplateAction);
                        break;
                    case ActionType.Inline:
                        type = typeof(InlineAction);
                        break;
                    case ActionType.Custom:
                        type = typeof(CustomAction);
                        break;
                    default:
                        throw new Exception("absurd!");
                }
                XSerializer ser = new XSerializer(type) { Context = context };
                this.Action = (Action)ser.Deserialize(actionElement);
            }
        }

        void IXSerializable.Save(IServiceProvider context, XElement element)
        {
            XHelper.Default.SaveByXAttributes(context, element, this);
            XNamespace ns = Consts.ScheduleNamespaceURI;

            if (this.Trigger != null)
            {
                XElement triggerElement = new XElement(ns + "trigger");
                element.Add(triggerElement);
                XSerializer ser = new XSerializer(this.Trigger.GetType()) { Context = context };
                triggerElement.SetAttributeValue("type", this.Trigger.Type.ToString());
                ser.Serialize(triggerElement, this.Trigger);
            }

            if (this.Action != null)
            {
                XElement actionElement = new XElement(ns + "action");
                element.Add(actionElement);
                XSerializer ser = new XSerializer(this.Action.GetType()) { Context = context };
                actionElement.SetAttributeValue("type", this.Action.Type.ToString());
                ser.Serialize(actionElement, this.Action);
            }

            if (this.CustomParams != null)
            {
                element.Add(XElement.Parse(this.CustomParams));
            }

        }

        #endregion
    }
}

File: branches/Stheno/Fantasy.Jobs/src/fantasy/jobs/scheduling/ScheduleItem.java
package fantasy.jobs.scheduling;

import fantasy.xserialization.*;

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
//[DataContract, KnownType(typeof(TimeTrigger)), KnownType(typeof(DailyTrigger)), KnownType(typeof(WeeklyTrigger)), KnownType(typeof(MonthlyTrigger)), KnownType(typeof(MonthlyDOWTrigger)), KnownType(typeof(TemplateAction)), KnownType(typeof(InlineAction)), KnownType(typeof(CustomAction)), XSerializable("schedule", NamespaceUri = Consts.ScheduleNamespaceURI)]
public class ScheduleItem extends IXSerializable
{
	public ScheduleItem()
	{

	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("name")]
	private String privateName;
	public final String getName()
	{
		return privateName;
	}
	public final void setName(String value)
	{
		privateName = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("author")]
	private String privateAuthor;
	public final String getAuthor()
	{
		return privateAuthor;
	}
	public final void setAuthor(String value)
	{
		privateAuthor = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("description")]
	private String privateDescription;
	public final String getDescription()
	{
		return privateDescription;
	}
	public final void setDescription(String value)
	{
		privateDescription = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("enabled")]
	private boolean privateEnabled;
	public final boolean getEnabled()
	{
		return privateEnabled;
	}
	public final void setEnabled(boolean value)
	{
		privateEnabled = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember]
	private Trigger privateTrigger;
	public final Trigger getTrigger()
	{
		return privateTrigger;
	}
	public final void setTrigger(Trigger value)
	{
		privateTrigger = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember]
	private Action privateAction;
	public final Action getAction()
	{
		return privateAction;
	}
	public final void setAction(Action value)
	{
		privateAction = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("priority")]
	private int privatePriority;
	public final int getPriority()
	{
		return privatePriority;
	}
	public final void setPriority(int value)
	{
		privatePriority = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("restartOnFailure")]
	private Restart privateRestartOnFailure;
	public final Restart getRestartOnFailure()
	{
		return privateRestartOnFailure;
	}
	public final void setRestartOnFailure(Restart value)
	{
		privateRestartOnFailure = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XArray(Name = "requiredUncPath"), XArrayItem(Name = "unc", java.lang.Class = typeof(UncPath))]
	private UncPath[] privateRequiredUncPths;
	public final UncPath[] getRequiredUncPths()
	{
		return privateRequiredUncPths;
	}
	public final void setRequiredUncPths(UncPath[] value)
	{
		privateRequiredUncPths = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("runOnlyIfIdle")]
	private boolean privateRunOnlyIfIdle;
	public final boolean getRunOnlyIfIdle()
	{
		return privateRunOnlyIfIdle;
	}
	public final void setRunOnlyIfIdle(boolean value)
	{
		privateRunOnlyIfIdle = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("startWhenAvailable")]
	private boolean privateStartWhenAvailable;
	public final boolean getStartWhenAvailable()
	{
		return privateStartWhenAvailable;
	}
	public final void setStartWhenAvailable(boolean value)
	{
		privateStartWhenAvailable = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember]
	private String privateCustomParams;
	public final String getCustomParams()
	{
		return privateCustomParams;
	}
	public final void setCustomParams(String value)
	{
		privateCustomParams = value;
	}


//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("multipleInstances")]
	private InstancesPolicy privateMultipleInstance = InstancesPolicy.forValue(0);
	public final InstancesPolicy getMultipleInstance()
	{
		return privateMultipleInstance;
	}
	public final void setMultipleInstance(InstancesPolicy value)
	{
		privateMultipleInstance = value;
	}


//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("expired")]
	private boolean privateExpired;
	public final boolean getExpired()
	{
		return privateExpired;
	}
	public final void setExpired(boolean value)
	{
		privateExpired = value;
	}

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[DataMember, XAttribute("deleteAfterExpired")]
	private boolean privateDeleteAfterExpired;
	public final boolean getDeleteAfterExpired()
	{
		return privateDeleteAfterExpired;
	}
	public final void setDeleteAfterExpired(boolean value)
	{
		privateDeleteAfterExpired = value;
	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#region IXSerializable Members

	private void Load(IServiceProvider context, XElement element)
	{
		XHelper.Default.LoadByXAttributes(context, element, this);

		XmlNamespaceManager nm = new XmlNamespaceManager(new NameTable());
		nm.AddNamespace("s", Consts.ScheduleNamespaceURI);

		XElement customParams = element.XPathSelectElement("s:custom", nm);
		if (customParams != null)
		{
			this.setCustomParams(customParams.toString());
		}


		XElement triggerElement = element.XPathSelectElement("s:trigger", nm);
		if (triggerElement != null)
		{
			java.lang.Class type;
			TriggerType triggerType = TriggerType.valueOf((String)triggerElement.Attribute("type"));
			switch (triggerType)
			{
				case Time:
					type = TimeTrigger.class;
					break;
				case Daily:
					type = DailyTrigger.class;
					break;
				case Weekly:
					type = WeeklyTrigger.class;
					break;
				case Monthly:
					type = MonthlyTrigger.class;
					break;
				case MonthlyDayOfWeek:
					type = MonthlyDOWTrigger.class;
					break;
				default:
					throw new RuntimeException("Absurd!");
			}
			XSerializer ser = new XSerializer(type);
			ser.Context = context;
			this.setTrigger((Trigger)ser.Deserialize(triggerElement));

		}

		XElement actionElement = element.XPathSelectElement("s:action", nm);
		if (actionElement != null)
		{
			java.lang.Class type;
			ActionType actionType = ActionType.valueOf((String)actionElement.Attribute("type"));
			switch (actionType)
			{
				case Template:
					type = TemplateAction.class;
					break;
				case Inline:
					type = InlineAction.class;
					break;
				case Custom:
					type = CustomAction.class;
					break;
				default:
					throw new RuntimeException("absurd!");
			}
			XSerializer tempVar = new XSerializer(type);
			tempVar.Context = context;
			XSerializer ser = tempVar;
			this.setAction((Action)ser.Deserialize(actionElement));
		}
	}

	private void Save(IServiceProvider context, XElement element)
	{
		XHelper.Default.SaveByXAttributes(context, element, this);
		XNamespace ns = Consts.ScheduleNamespaceURI;

		if (this.getTrigger() != null)
		{
			XElement triggerElement = new XElement(ns + "trigger");
			element.Add(triggerElement);
			XSerializer tempVar = new XSerializer(this.getTrigger().getClass());
			tempVar.Context = context;
			XSerializer ser = tempVar;
			triggerElement.SetAttributeValue("type", this.getTrigger().getType().toString());
			ser.Serialize(triggerElement, this.getTrigger());
		}

		if (this.getAction() != null)
		{
			XElement actionElement = new XElement(ns + "action");
			element.Add(actionElement);
			XSerializer tempVar2 = new XSerializer(this.getAction().getClass());
			tempVar2.Context = context;
			XSerializer ser = tempVar2;
			actionElement.SetAttributeValue("type", this.getAction().getType().toString());
			ser.Serialize(actionElement, this.getAction());
		}

		if (this.getCustomParams() != null)
		{
			element.Add(XElement.Parse(this.getCustomParams()));
		}

	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#endregion
}